File: LeetCode/LC-1020-Number-Of-Enclave/Solution.py
class Solution(object):
    cnt = 0
    def numEnclaves(self, grid):
        ROW , COL = len(grid) , len(grid[0])
        global cnt
        visited = set()
        def dfs(row , col):
            res = 0
            global cnt
            if row < 0 or col < 0 or row == ROW or col == COL:
                return 0

            if grid[row][col] == 0 or (row , col) in visited:
                return 1
            

            visited.add((row,col))

            res = min(
                dfs(row,col-1),
                dfs(row, col +1 ),
                dfs(row-1 , col),
                dfs(row+1,col)
            )

            if res != 0 :
                cnt+=1
            else:
                cnt = 0
            return res
        
        res = 0
        for row in range(ROW):
            for col in range(COL):
                cnt = 0
                if grid[row][col] == 1 and (row,col) not in visited:
                    dfs(row,col)
                    res += cnt
        return res

File: LeetCode/LC-1020-Number-Of-Enclave/Solution.java
class Solution {
    private Set<List<Integer>> visited ;
    private int cnt;
    private int ROW,COL;

    public static int min(int a , int b , int c , int d){
        int[] arr = {a,b,c,d};
        Arrays.sort(arr);
        return arr[0];
    }

    public int dfs(int[][] grid,int row , int col){
        int res = 0;
        if(row < 0 || row == ROW || col < 0 || col == COL) {
            return 0;
        }
        if(grid[row][col] == 0 || visited.contains(Arrays.asList(row,col)) ){
            return 1;
        }

        visited.add(Arrays.asList(row,col));

        res = min(
            dfs(grid,row, col-1),
            dfs(grid,row, col+1),
            dfs(grid,row-1,col),
            dfs(grid,row+1, col)
        );

        cnt = res!=0 ? cnt+1 : 0;

        return res;
    }

    public int numEnclaves(int[][] grid) {
        ROW = grid.length;COL = grid[0].length;
        visited = new HashSet();

        int res = 0;

        for( int row = 0 ;row < ROW ;row++){
            
            for(int col = 0; col < COL ;col++){
                cnt = 0;
                if(grid[row][col] == 1 && !visited.contains(Arrays.asList(row,col)) )
                dfs(grid,row,col);
                res += cnt;
            }
        }

        return res;
    }
}